gi-gtk-declarative/src/GI/Gtk/Declarative/Container/Grid.hs

```diff
@@ -0,0 +1,68 @@
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+{-# LANGUAGE DeriveFunctor         #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE NamedFieldPuns        #-}
+{-# LANGUAGE OverloadedLabels      #-}
+{-# LANGUAGE RecordWildCards       #-}
+
+-- | Implementation of 'Gtk.Grid' as a declarative container.
+module GI.Gtk.Declarative.Container.Grid
+  ( GridChild(..)
+  , GridChildProperties(..)
+  , defaultGridChildProperties
+  ) where
+
+import           Data.Default.Class                 (Default (def))
+import           Data.Int                           (Int32)
+import           Data.Vector                        (Vector)
+import           Data.Word                          (Word32)
+import qualified GI.Gtk                             as Gtk
+
+import           GI.Gtk.Declarative.Container.Class
+import           GI.Gtk.Declarative.EventSource
+import           GI.Gtk.Declarative.Patch
+import           GI.Gtk.Declarative.Widget
+
+-- | Describes a child widget to be added with to a 'Grid'.
+data GridChild event =
+  GridChild
+    { properties :: GridChildProperties
+    , child      :: Widget event
+    }
+  deriving (Functor)
+
+-- | Values used when /packing/ child widgets into grids.
+data GridChildProperties =
+  GridChildProperties
+    { height     :: Int32
+    , width      :: Int32
+    , leftAttach :: Int32
+    , topAttach  :: Int32
+    }
+
+-- | Defaults for 'GridChildProperties'. Use these and override
+-- specific fields.
+defaultGridChildProperties :: GridChildProperties
+defaultGridChildProperties =
+  GridChildProperties {height = 1, width = 1, leftAttach = 0, topAttach = 0}
+
+instance Default GridChildProperties where
+  def = defaultGridChildProperties
+
+instance Patchable GridChild where
+  create = create . child
+  patch s b1 b2 = patch s (child b1) (child b2)
+
+instance EventSource GridChild where
+  subscribe GridChild {..} = subscribe child
+
+instance ToChildren Gtk.Grid Vector GridChild
+
+instance IsContainer Gtk.Grid GridChild where
```

Comment:
I'm not entirely sure how this will behave when based on the default `IsContainer` patching behavior. IIRC, the patching algorithm is based on the order of passed-in child widgets, and assumes that they can be accessed by index in some linear structure.

I'm not sure it's a problem, but consider the following. Left and top attachment can be used to position grid children in an order _different from the order in which they are passed_ in the argument list. If the passed-in ordering is the same on each update, this shouldn't be a problem. But if, for some reason, the passed-in order varies even though the explicit positioning (using left-attach/top-attach) is the same, then patching would be inefficient. It wouldn't patch the same underlying widget on re-renders, and thus have to recreate widgets.

Again, I don't see this as a huge issue, it's more of a performance concern.

Reply:
Related to this, I was thinking that another way to represent a declarative grid widget could be by passing a list of lists of cells. Each cell could specify spanning width/height (instead of the default 1). This would be analogous to tables in HTML, with colspan/rowspan. It would couple the ordering of passed-in widgets to the rendered order in the grid. The implementation would calculate the underlying positions to be set in the GTK grid.

Reply:
Again, just some thoughts. I say we merge this now. If we do something else later, it can be another container widget.

Reply:
> Related to this, I was thinking that another way to represent a declarative grid widget could be by passing a list of lists of cells. Each cell could specify spanning width/height (instead of the default 1). This would be analogous to tables in HTML, with colspan/rowspan. It would couple the ordering of passed-in widgets to the rendered order in the grid. The implementation would calculate the underlying positions to be set in the GTK grid.

Interesting. I suppose this might be more user friendly, too - since it requires less juggling of `leftAttach`/`topAttach`. Some kind of spacer widget might be needed, though I suppose an empty `Label` would probably be fine.